### PR TITLE
music fixes and attempted memory optimization

### PR DESCRIPTION
**Pull request by zephyrside**
ye

### CODE REVIEW

**Comment by kandowontu** [comment]
Zenith is gonna check this out in ~90 mins!

**Comment by ZenithNeko** [comment]
hey I'm so super sorry I just got home and I feel horrendously motion sick, I'll need a little extra time to recover before I can get to this

**Comment by kandowontu** [comment]
Love the options redo, as long as we can keep it under 9 its great!

**Comment by zephyrside** [comment]
i tried optimizing some memory with the other two commits and accidentally broke the music and sfx (i clearly can't do assembly language)

**Comment by kandowontu** [comment]
Skill issue 😝

**Comment by kandowontu** [comment]
Have you tried clearing your sram? Go to options and clear default settings (or clear your save file) it’s probably conflicting with old sram save data

**Comment by kandowontu** [comment]
oh you clobbered A here

.proc _music_play
    ldx	_options
    txa
    and #$08
    tax
    beq	start
    rts

**Comment by kandowontu** [comment]
Fix: 
.proc _music_play
    pha
    ldx	_options
    txa
    and #$08
    tax
    beq	start
    pla
    rts
start:
    pla

**Comment by zephyrside** [comment]
ah. i see
thank you

**Comment by kandowontu** [comment]
now you clobbered Y and all the sfx are the same xD

Fix:
.proc _sfx_play
        tax
        jsr popa
	pha
	tya
	pha
        ldy	_options
        tya
        and #$10
        tay
        bne end
	pla
	tay
	pla
        jmp famistudio_sfx_play

    end:
	pla
	tay
	pla
        rts
.endproc

**Comment by zephyrside** [comment]
:agony:

File: SAUCE/famidash.h [hunk 1]
```diff
@@ -151,11 +151,21 @@ unsigned char LEVELCOMPLETE[LEVEL_COUNT];
...
+//unsigned char oneptwoplayer;
+//unsigned char jumpsound;
+//unsigned char platformer;
+//unsigned char musicoff;
+//unsigned char sfxoff;
+unsigned char options;
+#define oneptwoplayer 0x01
+#define jumpsound 0x02
+#define platformer 0x04
+#define musicoff 0x08
```
Comment (jroweboy) [inline SAUCE/famidash.h:165]: change musicoff and sfxoff to bit 7 and bit6 instead
```c
#define musicoff 0x80
#define sfxoff 0x40
```

File: LIB/nesdash.s [hunk 2]
```diff
@@ -976,12 +976,18 @@ ParallaxBufferCol5:
 .popseg
 
 ;void __fastcall__ music_play(unsigned char song);
-.import _musicoff
+.import _options
 .proc _music_play
-    ldx	_musicoff
-    beq	start
+    pha
```
Comment (jroweboy) [inline LIB/nesdash.s:981]: then you can use bit to test for musicoff like this
```asm
.proc _music_play
  bit _options ; sets carry to bit 7 of _options without affecting a
  bvc start
    rts
start:
  LDY #$00
```

File: LIB/nesdash.s [hunk 3]
```diff
@@ -1027,15 +1033,27 @@ music_counts:
...
+.import _options
 .proc _sfx_play
-        tax
-        jsr popa
-        ldy	_sfxoff
-        bne end
-        jmp famistudio_sfx_play
+    tax
+    jsr popa
+	pha
```
Comment (jroweboy) [inline LIB/nesdash.s:1040]: and do the same here to check for sfxoff
```asm
.proc _sfx_play
bit _options ; bit 6 is copied to the overflow flag
bvc play
  rts
play:
tax
jsr popa
jmp famistudio_sfx_play
```

**Review by jroweboy — CHANGES_REQUESTED** [review]

**Comment by zephyrside** [comment]
muchas gracias

**Comment by zephyrside** [comment]
wait lmao
invalid opcode moment
image

**Comment by jroweboy** [comment]
Edit: I forgot it sets the N flag not the C flag, so for music off it should be
```
bpl musicon
  rts
musicon:
```

File: LIB/nesdash.s [hunk 4]
```diff
@@ -976,14 +976,13 @@ ParallaxBufferCol5:
...
-start:
-    LDY #$00
-    TSX
+.import _options
+.proc _music_play  
+    bit _options ; sets carry to bit 7 of _options without affecting a  
+    bvc start  
+    rts  
+start:  
+    LDY #$00
```
Comment (jroweboy) [inline LIB/nesdash.s:985]: the `tsx` still needs to be here which is why you are crashing

File: LIB/nesdash.s [hunk 5]
```diff
@@ -976,14 +976,13 @@ ParallaxBufferCol5:
...
-    ldx	_musicoff
-    beq	start
-    rts
-start:
-    LDY #$00
-    TSX
+.import _options
+.proc _music_play  
+    bit _options ; sets carry to bit 7 of _options without affecting a  
+    bvc start
```
Comment (jroweboy) [inline LIB/nesdash.s:982]: i goofed. it should be `sets N flag` and use
```
bpl start
```
instead

**Review by jroweboy — CHANGES_REQUESTED** [review]

**Comment by zephyrside** [comment]
alright that should do it

File: LIB/nesdash.s [hunk 6]
```diff
@@ -1027,16 +1026,15 @@ music_counts:
...
-        tax
-        jsr popa
-        ldy	_sfxoff
-        bne end
-        jmp famistudio_sfx_play
-
-    end:
-        rts
+.import _options
+.proc _sfx_play
```
Comment (jroweboy) [inline LIB/nesdash.s:1030]: mmm bad news, one more thing. this needs to be after popa so like this
```asm
.proc _sfx_play
    tax
    jsr popa
    
    bit _options ; bit 6 is copied to the overflow flag
    bvc play
    rts
play:
    jmp famistudio_sfx_play 
.endproc
```

**Comment by ZenithNeko** [comment]
looks and sounds good, merging
apologies for the long wait, ended up passing out all evening and wasn't allowed on the pc (and then had school)

**Pull request merged**